README.md
[![foundry-shield]][foundry-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![All Release Downloads](https://img.shields.io/github/downloads/MajorVictory/RetroUI-P5e/total.svg)]()

<br />
<p align="center">
  <a href="https://github.com/MajorVictory/RetroUI-P5e">
    <img src="images/p5e-thumb.png" alt="Logo" width="128" height="128">
  </a>
  <h3 align="center">Retro UI (P5e) - UI overhaul module for FoundryVtt</h3>
  <p align="center">
    A module for FoundryVtt that overhauls the UI with a customizable style themed after early handheld console games.
    <br />
    <br />
    <a href="https://github.com/MajorVictory/RetroUI-P5e/issues">Report Bug / Request Feature</a>
  </p>
</p>

## Installation

### Requirements

**[Core Compatibility]**:
 * *FoundryVTT* 0.6.5-0.6.6  

**[Systems Compatibility]**:
 * *dnd5e* 0.95-0.96
 * *Simple World-Building* 0.36

**[Required Modules]**
 * None

**[Specific Module Compatibility]**
 * dnd5e - Alt5eSheet
 * dnd5e - Tidy5eSheet
 * DynamicEffects
 * itemcollection
 * The Furnace
 * Monsterblock
 * Simple Dice Roller
 * [Request More](https://github.com/MajorVictory/RetroUI-P5e/issues)

**[Known Issues]**
 * *Unknown Compatibility* with game specific sheets and dialogs

### Latest Release

You can download the module in it's current state by pasting the following manifest url into Foundry.

> https://raw.githubusercontent.com/MajorVictory/RetroUI-P5e/master/module.json

![](readme/Install-7-31-2020.png)

## Changelog

[Previous Versions Changelog](CHANGELOG.md)

## Screenshots 

### Character sheets

![](readme/Sheets-9-4-2020.png)

Monsterblock

![](readme/Monsterblock-9-8-2020.png)

Tidy5eSheet

![](readme/tidy5esheet-9-9-2020.png)

Simple Dice Roller

![](readme/SimpleDiceRoller-9-14-2020.png)

### Customizable Colors

![](readme/Customize-9-4-2020.png)

![](readme/Controldex-9-3-2020.png)

### All Vanilla Menus and Dialogs 

![](readme/Overview-9-4-2020.png)

![](readme/Chat-9-4-2020.png)

## License

Retro UI (P5e) is a module for Foundry VTT by Majorvictory and is licensed under a [Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).

This work is licensed under Foundry Virtual Tabletop [EULA - Limited License Agreement for module development from May 29, 2020](https://foundryvtt.com/article/license/).

[foundry-shield]: https://img.shields.io/badge/Foundry-v0.6.6-informational
[foundry-url]: https://foundryvtt.com/
[forks-shield]: https://img.shields.io/github/forks/MajorVictory/RetroUI-P5e.svg?style=flat-square
[forks-url]: https://github.com/MajorVictory/RetroUI-P5e/network/members
[stars-shield]: https://img.shields.io/github/stars/MajorVictory/RetroUI-P5e.svg?style=flat-square
[stars-url]: https://github.com/MajorVictory/RetroUI-P5e/stargazers
[issues-shield]: https://img.shields.io/github/issues/MajorVictory/RetroUI-P5e.svg?style=flat-square
[issues-url]: https://github.com/MajorVictory/RetroUI-P5e/issues

CHANGELOG.md
# Changelog

## Version 1.4.0

 * Rewritten core in preparation for conversion to Whetstone style.
 * Added new logo
 * Module Support: Simple Dice Roller (fixes #6)

## Version 1.3.2

 * System Support: Simple Worldbuilding

## Version 1.3.1

 * Context menus
 * Scrollbars
 * Encumbrance bars
 * Renamed master style includes version to prevent caching between updates

## Version 1.3.0

 * Module Support: Tidy5eSheet (fixes #2)
 * Module Support: Monsterblock

## Version 1.2.2

 * Module Support: fixes styling for backpack items from itemcollection module. (fixes #4)

## Version 1.2.1

 * Bugfix: macro textbox becomes unusable when Furnace Module enabled. (fixes #3)

## Version 1.2

 * Initial support for Alt5e Character sheets
 * Bugfix: Controldex button didn't always get created
 * Minor fixes for Firefox layouts

## Version 1.1

 * Intitial Release


p5e-styles-scripts.js

function p5eGetBGData(num) {
	let slice = '';
	switch(num) {
		case "01": slice = '16 round'; break;
		default: slice = '16 round';
	}
	return 'url(\'/modules/RetroUI-P5e/images/chatborders/'+num+'.png\') '+slice;
}

function p5eCreateColors(hex, name) {
	let result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);

    let r = parseInt(result[1], 16);
    let g = parseInt(result[2], 16);
    let b = parseInt(result[3], 16);

    r /= 255, g /= 255, b /= 255;
    let max = Math.max(r, g, b), min = Math.min(r, g, b);
    let h, s, l = (max + min) / 2;

    if(max == min){
        h = s = 0; // achromatic
    } else {
        let d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
        switch(max) {
            case r: h = (g - b) / d + (g < b ? 6 : 0); break;
            case g: h = (b - r) / d + 2; break;
            case b: h = (r - g) / d + 4; break;
        }
        h /= 6;
    }
    s = Math.round(s * 100);
    l = Math.round(l * 100);
    h = Math.round(h * 360);
    let a = 1;

    let colors = {
    	value: 'hsla(' + h + ', ' + s + '%, ' + l + '%, ' + a + ')',
    	quarter: 'hsla(' + h + ', ' + s + '%, ' + l + '%, 0.25)',
    	half: 'hsla(' + h + ', ' + s + '%, ' + l + '%, 0.5)',
    	threequarter: 'hsla(' + h + ', ' + s + '%, ' + l + '%, 0.75)',
    	shadow: 'hsla(' + h + ', ' + s + '%, 25%, ' + a + ')',
        dark: 'hsla(' + h + ', ' + s + '%, 25%, 0.5)',
        light: 'hsla(' + h + ', 100%, 50%, ' + a + ')',
        darker: 'hsla(' + h + ', ' + s + '%, ' + Math.max(0, (l - 10)) + '%, ' + a + ')',
        lighter: 'hsla(' + h + ', ' + s + '%, ' + Math.min(100, (l + 10)) + '%, ' + a + ')'
    }

    for (let colortype in colors) {
    	let value = colors[colortype];
    	let propname = '--p5e-'+name+(colortype != 'value' ? '-'+colortype : '');
    	document.documentElement.style.setProperty(propname, value);
    	
    }
}

Hooks.once('init', () => {

    game.settings.registerMenu("RetroUI-P5e", "RetroUI-P5e", {
        name: game.i18n.localize("P5ESTYLES.Config"),
        label: game.i18n.localize("P5ESTYLES.ConfigTitle"),
        hint: game.i18n.localize("P5ESTYLES.ConfigHint"),
        icon: "fas fa-paint-brush",
        type: RetroUIP5eStylesConfigDialog,
        restricted: false
    });

    game.settings.register("RetroUI-P5e", "settings", {
        name: game.i18n.localize("P5ESTYLES.Config"),
        scope: "client",
        default: RetroUIP5eStylesConfig.getDefaults,
        type: Object,
        config: false,
        onChange: settings => {
			RetroUIP5eStylesConfig.apply(settings);
        }
    });
});

Hooks.once('ready', () => {
    RetroUIP5eStylesConfig.apply(RetroUIP5eStylesConfig.getOptions);
});

Hooks.once('renderSceneControls', () => {
    // controldex button 1 - opens styles config menu
    $('#controls').append($('<p>', { class: 'dexbutton dexbutton1' }).click(function(event) {
            const menu = game.settings.menus.get("RetroUI-P5e.RetroUI-P5e");
            if ( !menu ) return ui.notifications.error("No submenu found for the provided key");
            const app = new menu.type();
            return app.render(true);
        })
    );
});

// main class to hold default configs and current settings
export class RetroUIP5eStylesConfig {

	static get getDefaults() {
		return {
    		controldexPresets: {
				'#f20c32': 'Red',
				'#0a2dc9': 'Blue',
				'#0ac92a': 'Green',
				'#f2eb0c': 'Yellow',
				'#f2a20c': 'Gold',
				'#bebebe': 'Silver',
				'#02e2ff': 'Crystal',
				'#91071e': 'Ruby',
				'#072091': 'Sapphire',
				'#ff7700': 'Fire Red',
				'#00fd27': 'Leaf Green',
				'#07911f': 'Emerald',
				'#979ac8': 'Diamond',
				'#ecb2c2': 'Pearl',
				'#959595': 'Platinum',
				'#d8c156': 'HeartGold',
				'#8c95b5': 'SoulSilver',
				'#373737': 'Black',
				'#ffffff': 'White',
				'#202020': 'Black 2',
				'#fbeaee': 'White 2',
				'#3280cf': 'X',
				'#ff465b': 'Y',
				'#d43400': 'Omega Ruby',
				'#3e6aa9': 'Alpha Sapphire',
				'#fe9539': 'Sun',
				'#53a1d9': 'Moon',
				'#e56638': 'Ultra Sun',
				'#1e7bd4': 'Ultra Moon',
				'#e6c300': 'Let\'s Go, Pikachu!',
				'#a88633': 'Let\'s Go, Eevee!',
				'#4dcdff': 'Sword',
				'#ff4ea1': 'Shield'
			},
    		sheetColorPresets: {
				"#fffbce": "Standard Tan",
	        	"#ff9d9d": "Trainer Red",
	        	"#e4a089": "Rocky Brown",
	        	"#f4ce88": "Mousy Yellow",
	        	"#68c098": "Misty Green",
				"#b5dc92": "Buggy Green",
	        	"#58d0d0": "Trainer Blue",
	        	"#c0a5da": "Trainer Purple",
                "#eee6dd": "Eggy Shell White",
                "#f0ace6": "Balloon Pink",
                "#c0c0c0": "Steely Grey",
                "#525252": "Edgy Black"
			},
    		chatBorders: {
				"01": "Chat Border #01",
		        "02": "Chat Border #02",
		        "03": "Chat Border #03",
		        "04": "Chat Border #04",
		        "05": "Chat Border #05",
		        "06": "Chat Border #06",
		        "07": "Chat Border #07",
				"08": "Chat Border #08",
		        "09": "Chat Border #09",
		        "10": "Chat Border #10",
		        "11": "Chat Border #11",
		        "12": "Chat Border #12",
		        "13": "Chat Border #13",
		        "14": "Chat Border #14",
				"15": "Chat Border #15",
		        "16": "Chat Border #16",
		        "17": "Chat Border #17",
		        "18": "Chat Border #18",
		        "19": "Chat Border #19",
		        "20": "Chat Border #20",
		        "21": "Chat Border #21",
		        "22": "Chat Border #22",
		        "23": "Chat Border #23",
		        "24": "Chat Border #24",
		        "25": "Chat Border #25",
		        "26": "Chat Border #26",
		        "27": "Chat Border #27",
		        "28": "Chat Border #28",
		        "29": "Chat Border #29"
			},
    		controldexColor: '#f20c32',
	    	SheetBackgroundColor: '#fffbce',
	    	SheetForegroundColor: '#58d0d0',
    		ChatBoxPublicBackgroundColor: '#58d0d0',
    		ChatBoxPublicBorder: '21',
    		ChatBoxWhisperBackgroundColor: '#c0a5da',
    		ChatBoxWhisperBorder: '03',
    		ChatBoxBlindBackgroundColor: '#ff9d9d',
    		ChatBoxBlindBorder: '11'
    	};
	}

	static get getOptions() {
		return mergeObject(RetroUIP5eStylesConfig.getDefaults, game.settings.get("RetroUI-P5e", "settings"));
	}

	static apply(options) {

		p5eCreateColors(options.controldexColor, 'bg-controldex-color');
		p5eCreateColors(options.SheetBackgroundColor, 'bg-color');
		p5eCreateColors(options.SheetForegroundColor, 'fg-color');
		p5eCreateColors(options.ChatBoxPublicBackgroundColor, 'chatbg-color-public');
		p5eCreateColors(options.ChatBoxWhisperBackgroundColor, 'chatbg-color-whisper');
		p5eCreateColors(options.ChatBoxBlindBackgroundColor, 'chatbg-color-blind');

		document.documentElement.style.setProperty('--p5e-chatborder-public', p5eGetBGData(options.ChatBoxPublicBorder));
		document.documentElement.style.setProperty('--p5e-chatborder-whisper', p5eGetBGData(options.ChatBoxWhisperBorder));
		document.documentElement.style.setProperty('--p5e-chatborder-blind', p5eGetBGData(options.ChatBoxBlindBorder));
	}
}

class RetroUIP5eStylesConfigDialog extends FormApplication {

	static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            title: game.i18n.localize("P5ESTYLES.Config"),
            id: "RetroUI-P5e-config",
            template: "modules/RetroUI-P5e/templates/settings.html",
            width: 510,
            height: 795,
            closeOnSubmit: true
        });
    }

    getData(options) {
    	let defaultOptions = RetroUIP5eStylesConfig.getDefaults;
    	let currentdata = game.settings.get('RetroUI-P5e', 'settings');
    	let returndata = this.reset ? defaultOptions : mergeObject(defaultOptions, currentdata);
    	return returndata;
    }

    activateListeners(html) {
        super.activateListeners(html);
        html.find('select[name="controldexColorPreset"],input[name="controldexColor"],input[name="controldexColorSelector"]').change(this.updateColor.bind(this));
        html.find('select[name="SheetBackgroundColorPreset"],input[name="SheetBackgroundColor"],input[name="SheetBackgroundColorSelector"]').change(this.updateColor.bind(this));
        html.find('select[name="SheetForegroundColorPreset"],input[name="SheetForegroundColor"],input[name="SheetForegroundColorSelector"]').change(this.updateColor.bind(this));
        html.find('select[name="ChatBoxPublicBackgroundColorPreset"],input[name="ChatBoxPublicBackgroundColor"],input[name="ChatBoxPublicBackgroundColorSelector"]').change(this.updateColor.bind(this));
        html.find('select[name="ChatBoxWhisperBackgroundColorPreset"],input[name="ChatBoxWhisperBackgroundColor"],input[name="ChatBoxWhisperBackgroundColorSelector"]').change(this.updateColor.bind(this));
        html.find('select[name="ChatBoxBlindBackgroundColorPreset"],input[name="ChatBoxBlindBackgroundColor"],input[name="ChatBoxBlindBackgroundColorSelector"]').change(this.updateColor.bind(this));
        html.find('input,select').change(this.onApply.bind(this));
        html.find('button[name="reset"]').click(this.onReset.bind(this));

        $('input[name="controldexColor"], input[name="SheetBackgroundColor"], input[name="SheetForegroundColor"], input[name="ChatBoxPublicBackgroundColor"], input[name="ChatBoxWhisperBackgroundColor"], input[name="ChatBoxBlindBackgroundColor"]').change();

        this.reset = false;
    }

    // updates 'borther' controls to all reflect the same values
    updateColor(event) {
    	let control = $(event.target);

    	// validate manual color
    	if (['controldexColor','SheetBackgroundColor', 'SheetForegroundColor','ChatBoxPublicBackgroundColor',
    		'ChatBoxWhisperBackgroundColor', 'ChatBoxBlindBackgroundColor'].includes(control.prop('name'))) {

    		let colortest = /^#[0-9A-F]{6}$/i.test(control.val());

    		if(!colortest) {
    			ui.notifications.warn(game.i18n.localize("P5ESTYLES.InvalidColorFormat"));
    			control.val('#000000');
    		}
    	}

    	let colorgroups = [
    		['controldexColorPreset', 'controldexColor', 'controldexColorSelector'],
    		['SheetBackgroundColorPreset', 'SheetBackgroundColor', 'SheetBackgroundColorSelector'],
    		['SheetForegroundColorPreset', 'SheetForegroundColor', 'SheetForegroundColorSelector'],
    		['ChatBoxPublicBackgroundColorPreset', 'ChatBoxPublicBackgroundColor', 'ChatBoxPublicBackgroundColorSelector'],
    		['ChatBoxWhisperBackgroundColorPreset', 'ChatBoxWhisperBackgroundColor', 'ChatBoxWhisperBackgroundColorSelector'],
    		['ChatBoxBlindBackgroundColorPreset', 'ChatBoxBlindBackgroundColor', 'ChatBoxBlindBackgroundColorSelector']
    	];

    	for (var group = colorgroups.length - 1; group >= 0; group--) {
    		let colorgroup = colorgroups[group];

    		if (colorgroup.includes(control.prop('name'))) {

    			for (var el = colorgroup.length - 1; el >= 0; el--) {
    				let brother = $('[name="'+colorgroup[el]+'"]');

    				//skip if setting myself
    				if (brother.name == control.prop('name')) continue;

					// no color matches a preset, set to custom
    				if (brother.prop('tagName') == 'SELECT') {
    					brother.val(control.val());
    					if (brother.val() == null) {
	    					brother.val('custom');
	    				}
    				} else if (brother.prop('tagName') == 'INPUT' && brother.prop('type') == 'text') {
						// if custom, set to black at first
    					if (control.val() == 'custom') {
	    					brother.val('#000000');
	    				} else {
	    					brother.val(control.val());
	    				}
    				} else {
						brother.val(control.val());
					}
    			}
    		}
    	}
    }

    onApply(event) {
    	event.preventDefault();
    }

    onReset() {
        this.reset = true;
        this.render();
    }

    async _updateObject(event, formData) {
        await game.settings.set('RetroUI-P5e', 'settings', formData);
        ui.notifications.info(game.i18n.localize("P5ESTYLES.SaveMessage"));
    }
}

RetroUI-P5e.js

Hooks.once('WhetstoneReady', () => {

    // register a default menu with Whetstone for configuration
    game.Whetstone.settings.registerMenu('RetroUI-P5e', 'RetroUI-P5e', {
        name: game.i18n.localize('P5ESTYLES.Config'),
        label: game.i18n.localize('P5ESTYLES.ConfigTitle'),
        hint: game.i18n.localize('P5ESTYLES.ConfigHint'),
        icon: 'fas fa-paint-brush',
        //type: '', // leave empty to use default WhetstoneThemeConfigDialog
        restricted: false
    });

    // register a theme
    game.Whetstone.themes.register('RetroUI-P5e', {

        // the following keys will be pulled from the module.json
        // name, title, description, version, author/authors, url

        // author/authors will be formattd to fit this format
        authors: [
            { name: 'MajorVictory', contact: 'Github', url: 'https://github.com/MajorVictory' }
        ],

        // will be merged with 'styles' entry in module.json
        // these sheets are considered to be the 'core' styles for this theme
        // these files are always loaded first when the theme is enabed
        // no system specific themes should be defined here
        styles: [
            'modules/RetroUI-P5e/styles/fonts/loadfonts.css',
            'modules/RetroUI-P5e/styles/RetroUI-P5e-1.4.0.css'
        ],

        // here you can specify system specific styles or optional styles
        substyles: {
            'dnd5e': {
                //sybstyle id
                name: 'dnd5e',
                // title used in menu
                title: 'P5ESTYLES.SystemNameDND5e',
                // hint displayed under title
                hint: 'P5ESTYLES.SheetOverrideNotednd5e',
                // (optional) specify system
                system: 'dnd5e',
                // (optional) specify system version minimum
                version: '0.96',
                // if true and system/version are specified, they must match current setup
                // if true and no system/version specified, acts as a global style on by default
                active: true,
                // list of stylesheets to activate with this substyle
                styles: [
                    'modules/RetroUI-P5e/styles/systems/dnd5e-0.96.css'
                ]
            },
            'worldbuilding': {
                name: 'worldbuilding',
                title: 'P5ESTYLES.SystemNameWorldBuilding',
                hint: 'P5ESTYLES.SheetOverrideNote',
                system: 'worldbuilding',
                version: '0.36',
                active: true,
                styles: [
                    'modules/RetroUI-P5e/styles/systems/worldbuilding.css'
                ]
            },
            'pf2e': {
                name: 'pf2e',
                title: 'P5ESTYLES.SystemNamePF2e',
                hint: 'P5ESTYLES.SheetOverrideNote',
                system: 'pf2e',
                version: '1.10.10.1973',
                active: true,
                styles: [
                    'modules/RetroUI-P5e/styles/systems/pf2e.css'
                ]
            },
            'pf2e-colors': {
                //this substyle will activate only on the pf2e system
                name: 'pf2e-colors',
                title: 'P5ESTYLES.PF2EColorOverride',
                hint: 'P5ESTYLES.PF2EColorOverrideNote',
                system: 'pf2e',
                active: true,
                styles: [
                    'modules/RetroUI-P5e/styles/systems/pf2e-sheetcolors.css'
                ]
            }
        },

        variables: [
            // extra types supported: 'shades', 'color'
            // shades: produces color variants with the specified color as a base
            // color: produces one color value only
            {
                name: '--p5e-bg-color',
                title: 'P5ESTYLES.SheetBackgroundColor',
                hint: '<hr>',
                value: '#fffbce', type: 'shades', presets: 'sheetColorPresets'
            },
            {
                name: '--p5e-fg-color',
                title: 'P5ESTYLES.SheetForeroundColor',
                hint: '<hr>',
                value: '#58d0d0', type: 'shades', presets: 'sheetColorPresets'
            },
            {
                name: '--p5e-bg-controldex-color',
                title: 'P5ESTYLES.ControldexColor',
                hint: '<hr>',
                value: '#f20c32', type: 'color', presets: 'controldexPresets'
            },
            {
                name: '--p5e-chatbg-color-public',
                title: 'P5ESTYLES.ChatBoxPublic',
                hint: 'P5ESTYLES.ChatBoxPublicHint',
                value: '#58d0d0', type: 'color', presets: 'sheetColorPresets'
            },
            {
                name: '--p5e-chatborder-public',
                title: 'P5ESTYLES.ChatBorderPublic',
                hint: '<hr>',
                value: "url('../images/chatborders/21.png') 16 round", type: String, presets: 'chatBorders'
            },
            {
                name: '--p5e-chatbg-color-whisper',
                title: 'P5ESTYLES.ChatBoxWhisper',
                hint: 'P5ESTYLES.ChatBoxWhisperHint',
                value: '#c0a5da', type: 'color', presets: 'sheetColorPresets'
            },
            {
                name: '--p5e-chatborder-whisper',
                title: 'P5ESTYLES.ChatBorderWhisper',
                hint: '<hr>',
                value: "url('../images/chatborders/03.png') 16 round", type: String, presets: 'chatBorders'
            },
            {
                name: '--p5e-chatbg-color-blind',
                title: 'P5ESTYLES.ChatBoxBlind',
                hint: 'P5ESTYLES.ChatBoxBlindHint',
                value: '#ff9d9d', type: 'color', presets: 'sheetColorPresets'
            },
            {
                name: '--p5e-chatborder-blind',
                title: 'P5ESTYLES.ChatBorderBlind',
                hint: '<hr>',
                value: "url('../images/chatborders/11.png') 16 round", type: String, presets: 'chatBorders'
            }
        ],

        presets: {
            controldexPresets: {
                '#f20c32': 'Red',
                '#0a2dc9': 'Blue',
                '#0ac92a': 'Green',
                '#f2eb0c': 'Yellow',
                '#f2a20c': 'Gold',
                '#bebebe': 'Silver',
                '#02e2ff': 'Crystal',
                '#91071e': 'Ruby',
                '#072091': 'Sapphire',
                '#ff7700': 'Fire Red',
                '#00fd27': 'Leaf Green',
                '#07911f': 'Emerald',
                '#979ac8': 'Diamond',
                '#ecb2c2': 'Pearl',
                '#959595': 'Platinum',
                '#d8c156': 'HeartGold',
                '#8c95b5': 'SoulSilver',
                '#373737': 'Black',
                '#ffffff': 'White',
                '#202020': 'Black 2',
                '#fbeaee': 'White 2',
                '#3280cf': 'X',
                '#ff465b': 'Y',
                '#d43400': 'Omega Ruby',
                '#3e6aa9': 'Alpha Sapphire',
                '#fe9539': 'Sun',
                '#53a1d9': 'Moon',
                '#e56638': 'Ultra Sun',
                '#1e7bd4': 'Ultra Moon',
                '#e6c300': 'Let\'s Go, Pikachu!',
                '#a88633': 'Let\'s Go, Eevee!',
                '#4dcdff': 'Sword',
                '#ff4ea1': 'Shield'
            },
            sheetColorPresets: {
                "#fffbce": "Standard Tan",
                "#ff9d9d": "Trainer Red",
                "#e4a089": "Rocky Brown",
                "#f4ce88": "Mousy Yellow",
                "#68c098": "Misty Green",
                "#b5dc92": "Buggy Green",
                "#58d0d0": "Trainer Blue",
                "#c0a5da": "Trainer Purple",
                "#eee6dd": "Eggy Shell White",
                "#f0ace6": "Balloon Pink",
                "#c0c0c0": "Steely Grey",
                "#525252": "Edgy Black"
            },
            chatBorders: {
                "url('../images/chatborders/01.png') 16 round": "Chat Border #01",
                "url('../images/chatborders/02.png') 16 round": "Chat Border #02",
                "url('../images/chatborders/03.png') 16 round": "Chat Border #03",
                "url('../images/chatborders/04.png') 16 round": "Chat Border #04",
                "url('../images/chatborders/05.png') 16 round": "Chat Border #05",
                "url('../images/chatborders/06.png') 16 round": "Chat Border #06",
                "url('../images/chatborders/07.png') 16 round": "Chat Border #07",
                "url('../images/chatborders/08.png') 16 round": "Chat Border #08",
                "url('../images/chatborders/09.png') 16 round": "Chat Border #09",
                "url('../images/chatborders/10.png') 16 round": "Chat Border #10",
                "url('../images/chatborders/11.png') 16 round": "Chat Border #11",
                "url('../images/chatborders/12.png') 16 round": "Chat Border #12",
                "url('../images/chatborders/13.png') 16 round": "Chat Border #13",
                "url('../images/chatborders/14.png') 16 round": "Chat Border #14",
                "url('../images/chatborders/15.png') 16 round": "Chat Border #15",
                "url('../images/chatborders/16.png') 16 round": "Chat Border #16",
                "url('../images/chatborders/17.png') 16 round": "Chat Border #17",
                "url('../images/chatborders/18.png') 16 round": "Chat Border #18",
                "url('../images/chatborders/19.png') 16 round": "Chat Border #19",
                "url('../images/chatborders/20.png') 16 round": "Chat Border #20",
                "url('../images/chatborders/21.png') 16 round": "Chat Border #21",
                "url('../images/chatborders/22.png') 16 round": "Chat Border #22",
                "url('../images/chatborders/23.png') 16 round": "Chat Border #23",
                "url('../images/chatborders/24.png') 16 round": "Chat Border #24",
                "url('../images/chatborders/25.png') 16 round": "Chat Border #25",
                "url('../images/chatborders/26.png') 16 round": "Chat Border #26",
                "url('../images/chatborders/27.png') 16 round": "Chat Border #27",
                "url('../images/chatborders/28.png') 16 round": "Chat Border #28",
                "url('../images/chatborders/29.png') 16 round": "Chat Border #29"
            }
        },

        // (optional) FormApplication extended class that creates the options dialog
        // if you register your own dialog menu you can specify it here
        // leaving this blank will use the Whetstone default config dialog
        // >                  This Value ---v   and   ---v  to this ---> 'RetroUI-P5e.RetroUI-P5e'
        // > game.settings.registerMenu('RetroUI-P5e', 'RetroUI-P5e', {...});
        dialog: '',

        // (optional) Thumbnail image
        img: 'modules/RetroUI-P5e/images/p5e-thumb.png',

        // (optional) Large preview image
        preview: 'modules/RetroUI-P5e/readme/Overview-9-4-2020.png',

        // (optional) Ensure the specified module is loaded before registering theme 
        dependencies: {
            'RetroUI-Core': ''
        },

        // (optional) known compatible systems/versions
        // if specified, sheets are attempted to be loaded in the following order
        // 1. /modules/<Module_Name>/styles/systems/<SystemID>.css
        // 2. /modules/<Module_Name>/styles/systems/<SystemID>-<Version>.css
        // 
        // <SystemID> : will be the current active system id
        // <Version> : will be the current active system's version
        // 
        // specified versions are used to indicate to the user if a theme may have issues with their world
        // These are NOT ENFORCED
        systems: {
            'core': '0.6.6',
            'dnd5e': '0.96',
            'worldbuilding': '0.36',
            'pf2e': '1.10.10.1973'
        },

        // (optional) known compatibilities
        compatible: {
            'alt5e': '1.3.4',
            'tidy5e-sheet': '0.2.21',
            'furnace':'',
            'dice-calculator': ''
        },

        // (optional) known conflicts
        conflicts: {
            'darksheet': '',
            'dnd-ui': ''
        },
    });
});

Hooks.on('renderSceneControls', () => {
    // controldex button 1 - opens styles config menu
    $('#controls').append($('<p>', { class: 'dexbutton dexbutton1' }).click(function(event) {
            const menu = game.Whetstone.settings.menus.get('RetroUI-P5e.RetroUI-P5e');
            if ( !menu ) return ui.notifications.error('No submenu found for the provided key');
            const app = new menu.type({theme:'RetroUI-P5e'});
            return app.render(true);
        })
    );
});